Fix meta_source_filenames bug and enable (meta)source_filenames appending of path and list

emiliom opened the pull request:
Addresses #678. More specifically, it enables `source_files_vars` in `prov.py` to handle the case of an input `source_paths` or `meta_source_paths` that includes both scalar elements (string or pathlib Path) and a list of such elements. This is the case that would arise with `combine_echodata` to append (combine) a converted file to a previously combined echodata object. Note that beyond the enhancement to handle `source_filenames` and `meta_source_filenames`, this PR does not address the larger, blocking issue described in #895, which currently makes that use case not possible.

I've done offline manual tests to confirm that the underlying `prov._source_files` function addresses that use case (heterogeneous list elements).

Also fixes the small bug described in #907

leewujung commented on echopype/utils/prov.py:
```diff
@@ -30,20 +35,63 @@ def echopype_prov_attrs(process_type: ProcessType) -> Dict[str, str]:
+
 def source_files_vars(
-    source_paths: Union[str, List[Any]], meta_source_paths: Union[str, List[Any]] = None
+    source_paths: Union[str, Path, List[Any], Tuple[Any], np.ndarray],
+    meta_source_paths: Union[None, str, Path, List[Any], Tuple[Any], np.ndarray] = None,
```
The `List[Any], Tuple[Any]` should actually be `List[str, Path], Tuple[str, Path]` as what you had in the docstring?

emiliom replied:
Heh, good eye. Initially I tried having `List[str, Path]` and `Tuple[str, Path]` in the argument signature. But it looks like type hinting doesn't allow that, b/c I got an error. So I went with `List[Any]` and `Tuple[Any]` but kept the more explicit version in the docstring in order to, well, be more explicit. I wonder if using `Union[str, Path]` instead of `str, Path` would do the trick. If it does, it'll make for an accurate but pretty long and hard-to-digest type specification. 

I also didn't try to see if `np.ndarray` can be followed by a sequence element type specification. I did some research just now, and it turns out that numpy added a typing library last year or so, [numpy.typing](https://numpy.org/devdocs/reference/typing.html). It includes a better replacement for `np.ndarray`: `numpy.typing.NDArray`. It also has `numpy.typing.ArrayLike`, which *may* be a more compact way to handle `List`, `Tuple` and `np.ndarray`. I'll look into it. It'll be great if that's the case!

emiliom replied:
After much R&D, I submitted a "perfect" (IMHO) implementation of the type hints using aliases, and it turns out it [fails with Python 3.8 ](https://github.com/OSOceanAcoustics/echopype/actions/runs/3744903528/jobs/6358775827#step:14:36) :sob: . I'll revisit it later.

leewujung commented on echopype/utils/prov.py:
```diff
@@ -30,20 +35,63 @@ def echopype_prov_attrs(process_type: ProcessType) -> Dict[str, str]:
     return prov_dict
 
 
+def _source_files(paths: Union[str, Path, List[Any], Tuple[Any], np.ndarray]):
```
How about renaming this to something like `_sanitize_source_list`? Thought it's good for "sanitize" to show up.

emiliom replied:
I like that suggestion. I'll implement it.

emiliom replied:
I've renamed the function to `_sanitize_source_files`. Though I just realized that your suggestion was `_sanitize_source_list`. How about we go with what I've done?

emiliom commented:
@leewujung Thanks for the review! I like your suggestion of adding a unit test. I could do it by tomorrow, if that still works with the release timeframe.

emiliom commented:
I've completed the individual modifications you brought up. The remaining task is creating a unit test.

emiliom commented:
Alright, I've added unit tests. @leewujung please take a look when you get a chance.

(FYI in the process I also discovered and fixed a small bug I had introduced in my last commit involving the handling of input in the form an np.ndarray)

leewujung commented:
@emiliom : thanks, the test looks good! Please feel free to merge.
